### PR TITLE
Added an async option that will write to I/O using a thread

### PR DESCRIPTION
**Pull request by thedrow**
This PR introduces a new option for I/O appenders that configures if the appender should use an auxiliary thread in order to write to the output.

This is useful (on average) when writing large buffers to I/O such as in the case of using the buffering appenders.
The option is turned off by default.

See benchmark.
Tests are not included because I need instructions on how to test this and where to place those tests.
The rolling file appender ignores this option for now because I'm not sure where to place the Thread.new code.

### CODE REVIEW

**Comment by thedrow** [comment]
I just noticed it's not writing. I have no idea why though.

**Comment by thedrow** [comment]
Hmm when I enable abort_on_exception I get:
```
/Users/omer.katz/Documents/logging/lib/logging/diagnostic_context.rb:445:in `initialize': can't create Thread: Resource temporarily unavailable (ThreadError)
	from /Users/omer.katz/Documents/logging/lib/logging/diagnostic_context.rb:445:in `new'
	from /Users/omer.katz/Documents/logging/lib/logging/diagnostic_context.rb:445:in `create_with_logging_context'
	from /Users/omer.katz/Documents/logging/lib/logging/diagnostic_context.rb:410:in `new'
	from /Users/omer.katz/Documents/logging/lib/logging/appenders/io.rb:79:in `canonical_write'
	from /Users/omer.katz/Documents/logging/lib/logging/appenders/buffering.rb:257:in `write'
	from /Users/omer.katz/Documents/logging/lib/logging/appender.rb:104:in `<<'
	from /Users/omer.katz/Documents/logging/lib/logging/logger.rb:167:in `block in <<'
	from /Users/omer.katz/Documents/logging/lib/logging/logger.rb:167:in `each'
	from /Users/omer.katz/Documents/logging/lib/logging/logger.rb:167:in `<<'
	from benchmark.rb:111:in `block (3 levels) in run'
	from benchmark.rb:111:in `times'
	from benchmark.rb:111:in `block (2 levels) in run'
	from /Users/omer.katz/.rvm/rubies/ruby-2.2.1/lib/ruby/2.2.0/benchmark.rb:288:in `measure'
	from /Users/omer.katz/.rvm/rubies/ruby-2.2.1/lib/ruby/2.2.0/benchmark.rb:370:in `item'
	from benchmark.rb:111:in `block in run'
	from /Users/omer.katz/.rvm/rubies/ruby-2.2.1/lib/ruby/2.2.0/benchmark.rb:172:in `benchmark'
	from /Users/omer.katz/.rvm/rubies/ruby-2.2.1/lib/ruby/2.2.0/benchmark.rb:204:in `bm'
	from benchmark.rb:109:in `run'
	from benchmark.rb:125:in `<main>'
```

**Comment by TwP** [comment]
This approach will conflict with the [periodic flushing](https://github.com/TwP/logging/blob/master/lib/logging/appenders/buffering.rb#L175-L195) available with the buffered appender module. There is also a race condition with this code. Two quick log messages in succession can create two new threads. Each of these threads will contend on the `syswrite` to the IO stream.

Take a look at the period flushing code. To enable asynchronous writes to the logging destination a new asynchronous logging decorator type will be needed. This code would be very similar to the buffering module - each log message is appended to a buffer, and then a single long-running thread would be responsible for the actual writes to the destination IO stream.

**Comment by thedrow** [comment]
I don't think we need two buffers both for periodic flushing and async writes. They should share the same buffer don't you think?

**Comment by TwP** [comment]
I think you can create asynchronous writes by using the buffering module with the `flush_period` set to zero. On each write the log message will be stored in the buffer, and a signal is sent to the flushing thread to write the contents of the buffer to the logging destination.

**Comment by thedrow** [comment]
@TwP But that's not what I'm after. I configured the logger to flush after 500 messages or when an error occurs. The flushing action as far as I can tell is a blocking action.
Writing 500 log messages shouldn't block the main thread or at least block the main thread as less as possible.

**Comment by thedrow** [comment]
See https://github.com/TwP/logging/blob/master/lib/logging/appenders/buffering.rb#L264
It might be preformed on the main thread if we hit the case where 500 messages have been written to the buffer but the flushing period hasn't been completed yet.

**Comment by TwP** [comment]
Okay, I think I'm understanding what you're after here. My apologies for taking so longer to get back to you as well.

tl;dr - this async flushing capability should be a feature of the `Buffering` appender module

> I think you can create asynchronous writes by using the buffering module with the `flush_period` set to zero.

In my view of the Logging universe here, this will write each message immediately using a separate thread. But your desired behavior is to buffer N messages and then write the buffer using a separate thread.

The `@periodic_flusher` should be renamed to `@async_flusher`. It will be responsible for flushing periodically (as it currently does) OR flushing asynchronously when the buffer reaches N messages. The reason I would like to move the asynchronous writing into the `Buffering` module is that we already have a thread in place. The implementation in this PR creates a new thread on every write. This will be costly in production systems, and it will cause extra GC churn.

I would love to hear your thoughts on the approach outlined here :thought_balloon:

**Comment by TwP** [comment]
I've taken a stab at this over in #135. Would love your thoughts on the approach outlined there. It should give you everything you need.

**Comment by TwP** [comment]
Closing this PR out in favor of #135
